---
title: "R Notebook"
output: html_notebook
---
#Exploratory analysis of single cell RNASeq data
Data from

##Filtration, normalization, and Dimensionality reduction/clustering analyses performed (PCA, t-SNE)

Import packages
```{r}
library(org.Hs.eg.db) 
library(Seurat)
library(scRNAseq)
library(dplyr) 
```

Functions
```{r}
#From list of marker hits (from Seurat) + list of known markers of immune genes, identify top 3 most likely cell types for a cluster
IdentifyLikelyCell<-function(immuneCells, cluster){
  cluster_cells<- immuneCells[rownames(immuneCells) %in% rownames(cluster),]
  CellMarkerCounts<- data.frame(table(cluster_cells$MaxCell))
  CellMarkerCounts<- CellMarkerCounts[order(CellMarkerCounts$Freq,decreasing=TRUE),]
  return(CellMarkerCounts[1:3,])
}
```

#Read in data, set up seurat object
```{r}
#Read in data- using bioconductor package "scRNASeq" for retrieving publically available scRNASeq datasets
wd<- "C:/Users/jjl71/OneDrive/Jenna/OneDrive/Desktop/Seurat/"

dataInput<-BaronPancreasData('human')
data<-as.data.frame(counts(dataInput))


#Set up seurat object
sdata<-CreateSeuratObject(counts=data, project="SC", min.cells=30)
sdata <- NormalizeData(sdata, verbose = FALSE)
```

#Visualize QC metrics, use for filtration
```{r}
VlnPlot(sdata, features = c("nFeature_RNA", "nCount_RNA"), ncol = 3)
```
Will want to remove outlier cells


```{r}
sdata <- subset(sdata, subset = nFeature_RNA > 100)
VlnPlot(sdata, features = c("nFeature_RNA", "nCount_RNA"), ncol = 3)
```

#Normalize
```{r}
sdata <- NormalizeData(sdata, normalization.method = "LogNormalize", scale.factor = 10000)
```

#Find variable features
```{r}
sdata <- FindVariableFeatures(sdata, selection.method = "vst", nfeatures = 1000)

#20 most variable features
top20 <- head(VariableFeatures(sdata), 20)

#Plot variable features, labeling top 20
plot <- VariableFeaturePlot(sdata)
LabelPoints(plot = plot, points = top20, repel = TRUE)

```

#Cell cycle: test for clustering based on likely cell cycle location of each cell
```{r}
#Read in list of cell cycle genes
cell_cycle <- read.csv(paste(wd, "CellCycleGenes.csv", sep=""))

#convert to gene symbols
cell_cycle$geneID <- unname(mapIds(org.Hs.eg.db, keys =cell_cycle$geneID, keytype = "ENSEMBL", column="SYMBOL"))

#Extract the G2/M genes
g2m_genes <- dplyr::filter(cell_cycle, phase == "G2/M") %>%
  pull(geneID) %>%
  as.character() 
  
#Extract the S genes
s_genes <- dplyr::filter(cell_cycle, phase == "S") %>%
  pull(geneID) %>%
  as.character() 

#Perform cell cycle scoring
sdata <- CellCycleScoring(sdata, g2m.features = g2m_genes, s.features = s_genes)

#Perform PCA and color by cell cycle phase
sdata = RunPCA(
  sdata,
  pc.genes = c(s_genes, g2m_genes),
  do.print = FALSE)

PCAPlot(sdata, 
        group.by= "Phase")
```
No clustering by cell cycle- don't need to include s.score and g2m.score in regression

#Dimensionality reduction
```{r}
#Apply linear transformation prior to PCA
sdata <- ScaleData(sdata, features = rownames(sdata))

#run pCA
sdata <- RunPCA(sdata, npcs = 30)

#plot PCA
DimPlot(sdata, reduction = "pca")
```
```{r}
#Plot heatmap of pcs
DimHeatmap(sdata, dims = 1:10, cells = 500, balanced = TRUE)
```
#Determine how many dimensions to use for further clustering
```{r}
#Using JackStraw resampling
sdata <- JackStraw(sdata, num.replicate = 100)
sdata <- ScoreJackStraw(sdata, dims = 1:20)

JackStrawPlot(sdata, dims = 1:20)
```
PCs 1-15 will be used for downstream clustering

#Perform additional clustering: UMAP and tSNE

#UMAP:
```{r}
sdata <- FindNeighbors(sdata, dims = 1:15)
sdata <- FindClusters(sdata, resolution = 0.5)

#Run UMAP clustering
sdata <- RunUMAP(sdata,  reduction = "pca", dims = 1:15)

DimPlot(sdata, reduction = "umap")
```
#tSNE clustering
```{r}
sdata <- RunTSNE(
  sdata,
  dims.use = 1:15,
  do.fast = TRUE)

#plot
DimPlot(sdata, reduction="tsne") 
```


#Find markers within each group
```{r}
# find all markers of cluster 1
cluster1.markers <- FindMarkers(sdata, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 5)
```

```{r}
# find all markers distinguishing cluster 1 from cluster 0 
cluster1.markers <- FindMarkers(sdata, ident.1 = 1, ident.2 = 0, min.pct = 0.25)
head(cluster1.markers, n = 5)

```

#Look at expression for top 5 genes DE between groups 1 and 2
```{r}
VlnPlot(sdata, rownames(head(cluster1.markers, n = 5)))
```
#Visualize expression of top 5 DE genes in cluster 1 in cluster plot
```{r}
FeaturePlot(sdata,rownames(head(cluster1.markers, n = 5)))
```

#Find markers in all groups
```{r}
# find markers for every cluster compared to all remaining cells, report only the positive ones (top 2)
sdata.markers <- FindAllMarkers(sdata, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
sdata.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```


#Attempt to classify clusters by overlap between their markers/markers of common immune cells

```{r}
#Read in cluster markers, previously analyzed and exported
sdata.markers<-read.csv(paste(wd, "AllMarkers.csv", sep="" ), row.names="X")

#Select only significant (adjusted p value) genes
markers.sig<- sdata.markers[sdata.markers$p_val_adj<=0.05,]

#Split into separate dfs for each cluster
markerList<- split(markers.sig, f= markers.sig$cluster )

#Read in df of immune cell markers. Convert to gene symbol, remove any NAs or duplicates
immuneCells<-read.csv(paste(wd, "mean_tpm_merged.csv", sep=""),stringsAsFactors = FALSE)
immuneCells$GeneSymbol <- mapIds(org.Hs.eg.db, keys = immuneCells$gene, keytype ="ENSEMBL", column="SYMBOL")
immuneCells<-na.omit(immuneCells)
immuneCells<-immuneCells[!(duplicated(immuneCells$GeneSymbol)),]
rownames(immuneCells)<-immuneCells$GeneSymbol
immuneCells<- immuneCells[,!(colnames(immuneCells) %in% c("gene", "GeneSymbol"))]

#Create df of "unique" markers: for each gene, only consider it a marker for the gene it's expressed highest in
uniqueMarkers<- as.data.frame(t(immuneCells))
uniqueMax<- apply(immuneCells, 1, which.max)
immuneCells$MaxCell<-colnames(immuneCells)[max.col(immuneCells)]

#Identify top 3 most likely cell types for each cluster
topCells<- lapply(markerList, IdentifyLikelyCell, immuneCells=immuneCells)

alltopCells<-topCells %>%  Reduce(function(dtf1,dtf2) full_join(dtf1,dtf2,by="Var1"), .)

#Manually look through top cells, assign each a most likely group. May be overlaps.
ClusterClassifications<- c("T.cell..CD8..naive..activated", "T.cell..CD4..naive..activated", "X","Monocyte..classical",
                           "NK.cell..CD56dim.CD16","Monocyte..non.classical", "X", "B.cell..naive",
                           "NK.cell..CD56dim.CD16","X", "X", "X",
                           "X", "Monocyte..non.classical" )
```

#Rename clusters based on markers- if known
```{r}
new.cluster.ids <- c("Cell type 1", "Cell type 2")
names(new.cluster.ids) <- levels(sdata)
sdata <- RenameIdents(sdata, new.cluster.ids)
DimPlot(sdata, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```